doc/compiler/optimizer/inlineFastPathLocations.md

```diff
@@ -0,0 +1,119 @@
+into a class property pointer which can be dereferenced again. All these kinds of dereferences cannot be written in normal
+Java code because they go through the JVM native data structures. They can be written in unsafe code. Unsafe path allows
+us to access memory without any safe type checking or bound checking, etc. It also allows us to look at native memory
+(`sun.misc.Unsafe.allocateMemory()`).
+
+Permission change is required to use unsafe APIs. It is not meant to be used in general programing. It is meant to be a
+tool to be used internally within JCL. Unsafe inlining is important because unsafe methods are used often by JCL. Calling
+into natives is slow. It is a lot faster to use JIT inlined expressions.
+
+`UnsafeFastPath` reduces unsafe calls to a "no control flow" sequence of IL trees to do the required operation. For example,
+`getInt(Object o, long offset)` adds the address of an object `o` and adds it with the offset and loads an integer. They
+can be expressed to a "no control flow" sequence of IL trees to do the required operation directly (e.g. `load` or `store`).
+
+The control flow that `UnsafeFastPath` avoids is what needs to be checked for determining which of the various cases the
+unsafe access falls in, e.g. the cases when we are accessing a static, array etc. It tries to avoid general cases and
+focuses on the specific cases/contexts where it knows enough about what the unsafe access is for to avoid that control flow
+in the pass.
+
+For example if an unsafe call is made in `java.util.concurrent.ConcurrentHashMap<K,V>`.  We could reason that the unsafe
+call in `ConcurrentHashMap` will never see a `NULL` array object. It will be guaranteed to be a reference array, non static.
+Things like this can be determined during the compilation. Therefore, all the checking can be skipped. We could just do
+the load or store.
+
+`UnsafeFastPath` also [runs as a part of](https://github.com/eclipse/omr/blob/e3a15a993c8aba80582aa1d6f3071e122acbd4c4/compiler/optimizer/OMROptimizer.cpp#L543) `ilgenStrategyOpts` before `recognizedCallTransformer`.
+It does a relatively simple substitution of certain calls very early without adding control flow before the inliner even
+runs (so it never sees these calls in the IL trees).
+
+Not all unsafe fast path is done in [UnsafeFastPath.cpp](https://github.com/eclipse-openj9/openj9/blob/master/runtime/compiler/optimizer/UnsafeFastPath.cpp). It is only used when there is no control flow. Inliner does the general unsafe inlining.
+It checks many conditions before it does unsafe inlining such as if it accesses an array or accesses a static variable,
+or if it is `NULL` or non `NULL`. There are various ways that these things can be checked.
+
+`UnsafeFastPath` is different from the other three optimizations. The other three optimizations are not aware of the
+context/caller. They do transformations in a general setting. `UnsafeFastPath` is more about the context where the unsafe
+call is in.
```

Comment:
"since it is the context that allows us to generate a no control flow sequence" may be worth adding

Reply:
addressed in 190c736